### PR TITLE
overlord: introduce InstanceKey to SnapState and SnapSetup, renames

### PR DESCRIPTION
**Pull request by bboozzoo**
Similarly to what we did to `snap.Info`, the PR introduces instance key to `SnapState` and `SnapSetup`. `SnapSetup.Name()` is renamed to `.SnapName()` and `.InstanceName()` to follow `snap.Info` naming pattern.

### CODE REVIEW

File: overlord/snapstate/snapmgr.go [hunk 1]
```diff
@@ -259,7 +271,15 @@ func (snapst *SnapState) CurrentInfo() (*snap.Info, error) {
...
 		return nil, ErrNoCurrent
 	}
-	return readInfo(cur.RealName, cur, 0)
+
+	name := snap.InstanceName(cur.RealName, snapst.InstanceKey)
+	info, err := readInfo(name, cur, 0)
+	if err != nil {
+		return nil, err
+	}
+	info.InstanceKey = snapst.InstanceKey
```
Comment (stolowski) [inline overlord/snapstate/snapmgr.go]: I might be missing a bigger picture, but looking at `readInfo` we call `snap.ReadInfo` and it sets `InstanceKey` there. Are we deliberately overwriting this here?

Comment (bboozzoo) [reply overlord/snapstate/snapmgr.go]: Thanks for spotting this. The pieces in `snap` are part of the same branch but were landed independently. Wil push a fix.

File: overlord/assertstate/assertmgr.go [hunk 2]
```diff
@@ -119,17 +119,17 @@ func doValidateSnap(t *state.Task, _ *tomb.Tomb) error {
...
+			return fmt.Errorf("cannot find supported signatures to verify snap %q and its hash (%v)", snapsup.InstanceName(), notFound)
 		}
 	}
 	if err != nil {
 		return err
 	}
 
 	db := DB(t.State())
-	err = snapasserts.CrossCheck(snapsup.Name(), sha3_384, snapSize, snapsup.SideInfo, db)
+	err = snapasserts.CrossCheck(snapsup.InstanceName(), sha3_384, snapSize, snapsup.SideInfo, db)
```
Comment (pedronis) [inline overlord/assertstate/assertmgr.go]: that value will be compared with the snap-declaration snap-name,  so we need to pass SnapName (but then errors from that function will use it)  or we need to teach CrossCheck what to do about instance names

File: overlord/ifacestate/handlers.go [hunk 3]
```diff
@@ -165,7 +165,7 @@ func (m *InterfaceManager) doRemoveProfiles(task *state.Task, tomb *tomb.Tomb) e
 	if err != nil {
 		return err
 	}
-	snapName := snapSetup.Name()
+	snapName := snapSetup.InstanceName()
```
Comment (pedronis) [inline overlord/ifacestate/handlers.go]: should we start renaming this kind of local vars to instanceName, later, or keep them snapName ?

File: overlord/snapstate/snapmgr.go [hunk 4]
```diff
@@ -77,9 +77,17 @@ type SnapSetup struct {
 
 	DownloadInfo *snap.DownloadInfo `json:"download-info,omitempty"`
 	SideInfo     *snap.SideInfo     `json:"side-info,omitempty"`
+
+	// InstanceKey is set by the user during installation and differs for
+	// each instance of given snap
+	InstanceKey string `json:"instance-key,omitempty"`
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: there is no non-test code setting this yet, is that right?

File: overlord/snapstate/snapmgr.go [hunk 5]
```diff
@@ -120,6 +128,10 @@ type SnapState struct {
 
 	// UserID of the user requesting the install
 	UserID int `json:"user-id,omitempty"`
+
+	// InstanceKey is set by the user during installation and differs for
+	// each instance of given snap
+	InstanceKey string `json:"instance-key,omitempty"`
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: same Q: there is no non-test code setting this yet, is that right?

File: overlord/snapstate/snapmgr.go (hunk 4 above)
Comment (bboozzoo) [reply overlord/snapstate/snapmgr.go]: Correct. The non-test code that makes use of InstanceKey in snapsup/snapst is currently only here https://github.com/bboozzoo/snapd/commits/bboozzoo/parallel-install-overlord

File: overlord/ifacestate/handlers.go (hunk 3 above)
Comment (bboozzoo) [reply overlord/ifacestate/handlers.go]: I can do some drive by changes, but rather in a separate PR carrying the bulk of changes in `overlord/snapstate`

File: overlord/assertstate/assertmgr.go (hunk 2 above)
Comment (mvo5) [reply overlord/assertstate/assertmgr.go]: @bboozzoo lets make sure we also have a test for this (either spread or unit) to make sure we never regress here.

**Pull request merged**